File: 0491-non-decreasing-subsequences/0491-non-decreasing-subsequences.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    int n;
    private void helper(int index, int[] nums, List<Integer> curr) {
        n = nums.length;
        if (curr.size() >= 2) {
            ans.add(new ArrayList<>(curr));
        }
        HashSet<Integer> st = new HashSet<>();
        for (int i = index; i < n; i++) {
            if (!curr.isEmpty() && curr.get(curr.size() - 1) > nums[i]) continue;
            if (st.contains(nums[i])) continue;
            st.add(nums[i]);

            curr.add(nums[i]);
            helper(i+1, nums, curr);
            curr.remove(curr.size()-1);
        }
    }
    public List<List<Integer>> findSubsequences(int[] nums) {
        n = nums.length;
        helper(0, nums, new ArrayList<>());
        return ans;
    }
}

File: 0491-non-decreasing-subsequences/0491-non-decreasing-subsequences.cpp
class Solution {
public:
    vector<vector<int>> ans;
    int n;
    void helper(int index, vector<int> nums, vector<int> curr) {
        if (curr.size() >= 2) {
            ans.push_back(curr);
        }
        unordered_set<int> st;
        for (int i = index; i < n; i++) {
            if (!curr.empty() && curr.back() > nums[i]) continue; // Checks if element breaks the root condition of the question, non-decreasing sequence
            if (st.count(nums[i])) continue; // Checks if the set has the element in it before hand
            st.insert(nums[i]);

            curr.push_back(nums[i]);
            helper(i + 1, nums, curr);
            curr.pop_back();
        }

    }
    vector<vector<int>> findSubsequences(vector<int>& nums) {
        n = nums.size();
        vector<int> curr;
        helper(0, nums, curr);
        return ans;
    }
};